### PR TITLE
`Programming exercises`: Add possibility to configure Docker run arguments on Jenkins

### LINKED ISSUE #6042: `Programming exercises`: Allow to define default arguments passed to Docker running the CI builds for automated tests
### Is your feature request related to a problem?

See #5971 for the base-implementation for Bamboo setups.

### Describe the solution you'd like

Jenkins setups should also support this feature.

The `args` option in the Jenkins pipeline script has to be filled accordingly. https://www.jenkins.io/doc/book/pipeline/docker/#caching-data-for-containers
The current variable replacement mechanism should be extendable to support that.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by b-fein**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
Closes #6042.
See #5971 for the Bamboo implementation.

### Description
Jenkins allows us to pass arguments to the Docker run command as well.
The YAML configuration format has been adapted a bit to be more flexible:
- short arguments like `-v host_path:/container_path` are now possible as well
- the same argument can be passed multiple times, e.g., `--env SOME_VAR=1 --env OTHER_VAR=2`

For Bamboo this change now means that the extra arguments are not passed as
```
--cpus=2
--memory=2g
--pids-limit=1000
--memory-swap=2g
```
but as
```
--cpus
"2"
--memory
"2g"
--pids-limit
"1000"
--memory-swap
"2g"
```
The two variants should be equivalent. So I assume this should not matter for the functionality.
I did not test it yet on Bamboo, since the test servers are not available at the moment.

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Create a new programming exercise.
2. The exercise details page you are on now contains a link to the build plan of the base and solution build plans below the task description.
3. Follow one of the links to Jenkins.
4. In the sidebar on the left select the ‘Configure’ entry.
5. Scroll down to the pipeline script definition (very bottom). The script should contain something like
```groovy
agent {
    docker {
        image '[…]'
        label 'docker'
        args '--cpus 2 […]'
    }
}
```

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| ProgrammingExerciseConfiguration.java | 100% of changed code | 100% of changed code |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java [hunk 1]
```diff
@@ -521,7 +521,6 @@ private DockerConfiguration dockerConfigurationFor(ProgrammingLanguage programmi
      * @return An array of string containing all the configured docker run argument key-value pairs prefixed with two dashes
      */
     private String[] getDefaultDockerRunArguments() {
-        // Bamboo needs all docker run arguments in separate lines
-        return programmingLanguageConfiguration.getDefaultDockerFlags().entrySet().stream().map(entry -> "--" + entry.getKey() + "=" + entry.getValue()).toArray(String[]::new);
+        return programmingLanguageConfiguration.getDefaultDockerFlags().toArray(new String[0]);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java:524]: What's the difference between `toArray(String[]::new)` and `toArray(new String[0])`, if any?

Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java:524]: The previous version was the `toArray` for a `Stream`, so it required a method reference. This is `List::toArray` which takes an actual array.

Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/bamboo/BambooBuildPlanService.java:524]: oh, I was not aware that there are two versions of this, expected this to be overloaded. Thanks.

**Review by Strohgelaender — DISMISSED** [review]
code 👍

**Review by grochlitzzz — COMMENTED** [review]
In Bamboo (ts1), it's still being passed like this. Jenkins server wasn't available for testing.
image

**Comment by grochlitzzz** [comment]
> 

> > In Bamboo (ts1), it's still being passed like this. Jenkins server wasn't available for testing.
> 
> @grochlitzzz Thanks for testing. Are you sure the release for this branch got deployed? According to the [Bamboo page](https://bamboo.ase.in.tum.de/deploy/viewDeploymentVersion.action?versionId=591855751) I don’t think it was.

That's weird, but thanks for the remark. My internet connection was unstable, maybe it didn't deploy and I didn't realize. Let me try again.

**Review by grochlitzzz — DISMISSED** [review]
Tested on Bamboo (ts1) and worked fine
image

**Comment by 4ludwig4** [comment]
:+1: manual test on local Jenkins/GitLab

Looks right in build job:
```
        agent {
            docker {
                image 'ls1tum/artemis-maven-template:java17-12'
                label 'docker'
                args '--cpus "2" --memory "2g" --memory-swap "2g" --pids-limit "1000"'
            }
        }
```
Gets passed called correctly in Build pipeline:
```
$ docker run -t -d -u 0:0 --cpus 2 --memory 2g --memory-swap 2g --pids-limit 1000 -w /var/jenkins_home/workspace/TESTCOURSETEST/TESTCOURSETEST-BASE --volumes-from 6eee6c2408e648f71d09f7ed2097745bcea1ea75f9205d4c0223fe22f3decc10 -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** ls1tum/artemis-maven-template:java17-12 cat
```

Changed the default parameters in the config (just cpus: 3) and it still works as intended:
```
        agent {
            docker {
                image 'ls1tum/artemis-maven-template:java17-12'
                label 'docker'
                args '--cpus "3"'
```

@b-fein: I'll also test Atlassian and changing the values if that's desired, as this can't be done on the testservers?

**Comment by b-fein** [comment]
@4ludwig4 Thanks for testing!

> I'll also test Atlassian and changing the values if that's desired, as this can't be done on the testservers?

I don’t think that’s necessary. If it’s not broken with the new default config like in the comment posted above yours, then it should work for changed values, too.

**Review by 4ludwig4 — DISMISSED** [review]

File: src/test/java/de/tum/in/www1/artemis/config/ProgrammingLanguageConfigurationTest.java [hunk 2]
```diff
@@ -91,6 +91,15 @@ void testOverriddenImageConfigurationSwift() {
         testOverriddenImageConfiguration(ProgrammingLanguage.SWIFT, List.of(ProjectType.XCODE));
     }
 
+    @Test
+    void testDockerRunFlags() {
+        final var config = new ProgrammingLanguageConfiguration();
+        config.setDefaultDockerFlags(getDockerFlags());
+
+        final List<String> expectedFlags = List.of("-v", "\"(pwd):/working_dir\"", "--cpus", "\"2\"", "--env", "\"VAR='value with spaces'\"", "--env", "\"SECOND_VAR=42\"");
```
Comment (dfuchss) [inline src/test/java/de/tum/in/www1/artemis/config/ProgrammingLanguageConfigurationTest.java:99]: Do you mean `$(pwd)` ?

File: src/test/java/de/tum/in/www1/artemis/config/ProgrammingLanguageConfigurationTest.java [hunk 3]
```diff
@@ -118,4 +127,9 @@ private Map<String, Map<String, String>> getDefaultImages() {
 
         return images;
     }
+
+    private List<ProgrammingLanguageConfiguration.DockerFlag> getDockerFlags() {
+        return List.of(new ProgrammingLanguageConfiguration.DockerFlag("-v", "(pwd):/working_dir"), new ProgrammingLanguageConfiguration.DockerFlag("--cpus", "2"),
```
Comment (dfuchss) [inline src/test/java/de/tum/in/www1/artemis/config/ProgrammingLanguageConfigurationTest.java:132]: Do you mean `$(pwd)` ?

**Review by dfuchss — COMMENTED** [review]
Code lgtm. I'll test asap. Just one minor remark

File: src/test/java/de/tum/in/www1/artemis/config/ProgrammingLanguageConfigurationTest.java (hunk 2 above)
Comment (b-fein) [reply src/test/java/de/tum/in/www1/artemis/config/ProgrammingLanguageConfigurationTest.java]: Sure, thanks. Changed it.

**Review by 4ludwig4 — APPROVED** [review]
